### ISSUE TITLE
Scala.js CollectionsTestBase.scala & descendents need to be ported to Scala Native.

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#1920]
The SN world is in flux now, with the pending Scala 2.12/2.13 changes and SN 0.next release.  So this is
a marker, not intended to distract core developers.

I believe the overarching vision is that the Scala Native test environment come, over iterations, the
more mature and capable Scala.js test environment.

I am currently working on providing TreeMap.scala, based on Scala.js RedBlackTree.scala and TreeSet.scala.
I believe this will make the scala-java-time tests run on SN with sufficient efficiency that devos will actually run them.
A workaround exist which will be useful for the next 6 to 9 months, but it is sub-optimal. Besides, who does not
like a TreeMap?
 
As part of this work I have a JUnit TreeMapTest.scala test.  I have large parts of a straight line test
running.  Looks like I could get more extensive coverage, and be more likely to pass PR review, if I 
wrote for Scala.js CollectionsOnMapsTest.scala.   Unfortunately, that file and the files it requires
have not yet been ported to SN.

This is the "very nice to have" bordering on "need" that I have, hence this Issue.

**Comment by LeeTibbert** [comment]
I do not want to impede or entangle current development work. I do think some preliminary design sketches
 can be done and submitted as Work In Progress (WIP) PRs, with the idea of being considered after 0.next settles.

At the appropriate time, the ideas in this entry can be reviewed, they take some brain power,
which, understandably, may not be available in the push to 0.next.

If encouraged, I can then proceed with personal sandbox work, leading to a WIP PR.
I have _plenty_ of other work to keep me busy & amused  until this proposal;s time rolls around.
My motivation is not just personal utility but the idea that other developers may encounter
the same issue. 

I have written for the Scala.js test environment when I submitted a PR there awhile ago.  Just now I took
a more detailed look at both it and the related current SN bits (AssertThrows.scala).  I do not
want to invest time in a direction which is either whacked or against some larger Plan.

The pending Scala 2.12 PR currently has no changes related to this discussion.

I think the first bar is consensus on overall design.  Then, I can invest time in a trial Work In Progress so 
that problems with that design can be discovered and people can see if/how the whole things work.

Because it is often easier for people to respond to a proposal rather than a blank sheet of paper,  let
me make a proposal on design.   I am open to and encourage discussion and alternate proposals.

   0) The intent at this pass is to build the SN CollectionsTestBase environment in parallel with existing code and
        not disturb it.  New code could use this environment and old code could be ported at leisure.

   1) I am somewhat leery of introducing duplication of files which then need to be kept synchronized, at some interval.  
       At some point down the road, it would be good to have a Common Point of Truth, perhaps under portable-scala.
       For now,  a trial SN implementation, even at the cost of cut & paste, would identify issues, if any, specific to Scala Native.

   2) I think my first attempt would be to duplicate the directory tree Scala.js 
       `project-root/test-suite/shared/src/test/scala/org/scalajs/testsuite/` as SN 
       `project-root/unit-tests/src/test/scala/org/scalaj/testsuite'.
    
    3) It is somewhat of a misdirection, but I would leave the package names as is:
        `org.scalajs.testsuite.javalib.util`. note the `org.scalajs`.   I think this will
        reduce the number of changes from the Scala.js reference copy. 

        Making those changes the first time is the cost of change. Making them
         on every re-port or back-port is a cost that can be avoided.

     3a) If the above hangs together, the current `AssertThrows.scala` would be
           superseded by its Scala.js original. Imports using it would need to be
           adjusted. I believe there should be no functional change.

**Comment by sjrd** [comment]
Your proposed plan sounds good, overall.

I am a bit on the fence regarding the package name for the tests. I wonder if keeping the `org.scalajs.*` package names makes sense. I guess that if the files are 100% copy-pasted, then that it is probably worth keeping the Scala.js package names.

**Comment by LeeTibbert** [comment]
Sjrd,  thank you for the timely and encouraging response.   I will proceed and see where it leads me/us.

I think the package name is one of those could-be-A-or-could-be-B design elements where a limited-effort trial 
implementation followed by a "run it up the flagpole" WIP may supply evidence to support one alternative or the
other.  

re: 100% copy pasted.  In a previous PR, you asked that I remove Scala.js copyright.  Should I do that here? 
     I think I should leave it but I do not want to introduce inconsistencies.  This is all '"penciled in" work proposed 
     anyway.

**Comment by WojciechMazur** [comment]
Already ported
